server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingMetadata.java

```diff
@@ -0,0 +1,215 @@
+ *
+ * When all source shards have transitioned to `DONE`, the resize is complete and this metadata may be removed from cluster state.
+ * We only allow at most a single resharding operation to be in flight for an index, so removing this metadata is a prerequisite
+ * to beginning another resharding operation.
+ */
+public class IndexReshardingMetadata implements ToXContentFragment, Writeable {
+    // ideally this would be final but the parser can't currently set it at construction time
+    private IndexReshardingState state;
+
+    private static final ParseField SPLIT_FIELD = new ParseField("split");
+    // This exists only so that tests can verify that IndexReshardingMetadata supports more than one kind of operation.
+    // It can be removed when we have defined a second real operation, such as shrink.
+    private static final ParseField NOOP_FIELD = new ParseField("noop");
+
+    private static final ObjectParser<IndexReshardingMetadata, Void> PARSER = new ObjectParser<>(
+        "index_resharding_metadata",
+        IndexReshardingMetadata::new
+    );
+
+    static {
+        PARSER.declareObjectOrNull(
+            IndexReshardingMetadata::setState,
+            (parser, c) -> IndexReshardingState.Split.fromXContent(parser),
+            null,
+            SPLIT_FIELD
+        );
+        PARSER.declareObjectOrNull(
+            IndexReshardingMetadata::setState,
+            (parser, c) -> IndexReshardingState.Noop.fromXContent(parser),
+            null,
+            NOOP_FIELD
+        );
+        PARSER.declareExclusiveFieldSet(SPLIT_FIELD.getPreferredName(), NOOP_FIELD.getPreferredName());
+        PARSER.declareRequiredFieldSet(SPLIT_FIELD.getPreferredName(), NOOP_FIELD.getPreferredName());
+    }
+
+    private IndexReshardingMetadata() {}
+
+    private void setState(IndexReshardingState state) {
+        this.state = state;
+    }
+
+    // for testing
+    IndexReshardingState getState() {
+        return state;
+    }
+
+    // visible for testing
+    IndexReshardingMetadata(IndexReshardingState state) {
+        this.state = state;
+    }
+
+    static IndexReshardingMetadata fromXContent(XContentParser parser) throws IOException {
+        return PARSER.parse(parser, null);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        String name = switch (state) {
+            case IndexReshardingState.Noop ignored -> NOOP_FIELD.getPreferredName();
+            case IndexReshardingState.Split ignored -> SPLIT_FIELD.getPreferredName();
+        };
+        builder.startObject(name);
+        state.toXContent(builder, params);
+        builder.endObject();
+
+        return builder;
+    }
+
+    public IndexReshardingMetadata(StreamInput in) throws IOException {
+        this();
+        var stateName = in.readString();
+
+        if (stateName.equals(SPLIT_FIELD.getPreferredName())) {
```

Comment:
I have a slight preference for switch here - just to keep it similar to `writeTo` mostly, but also a bit shorter I think. Really nit-picking, same order would also be good ;-).

Reply:
this was an if only because of the `getPreferredName` business (can't write `case SPLIT_FIELD.getPreferredName():`). I'll just create consts for the bare strings as well.

Reply:
Done in 0e69a76